domain/cached_repo.py

```diff
@@ -0,0 +1,40 @@
+import logging
+import threading
+from copy import deepcopy
+from datetime import datetime
+
+from domain.model.mesh_results import MeshResults
+from domain.repo import Repo
+
+logger = logging.getLogger(__name__)
+
+
+class CachedRepo:
+    """ CachedRepo allows for thread-safe, periodic updates of the repo data """
+
+    NEVER_UPDATED = datetime(1970, 1, 1)
+
+    def __init__(self, source_repo: Repo, monitored_test_id: str) -> None:
+        self._source_repo = source_repo
+        self._test_id = monitored_test_id
+        self._test_results = MeshResults()
+        self._update_timestamp = CachedRepo.NEVER_UPDATED
+        self._repo_access_lock = threading.Lock()
+
+    def update(self, lookback_minutes: int) -> None:
+        """ update is intended to be called periodically from an update thread """
+
+        try:
+            results = self._source_repo.get_mesh_test_results(self._test_id, lookback_minutes)
+            with self._repo_access_lock:
+                self._test_results = results
+            self._update_timestamp = datetime.now()
+        except Exception as err:
+            raise Exception(f"error updating repository data: {err}")
+
+    def data_timestamp(self) -> datetime:
+        return self._update_timestamp
+
+    def get_mesh_test_results(self) -> MeshResults:
+        with self._repo_access_lock:
+            return deepcopy(self._test_results)
```

Comment:
(Thought) Your approach is OK, but it works only for threads, which is not a showstopper in this case, because the app is mostly I/O bound, so Python threading works reasonably well (despite the GIL). However, have you considered dumping the JSON returned by the API to disk a reading it in the thread/process serving the page? To avoid race conditions, you'd need to create placeholder file e.g. `pending` for the data just being retrieved and rename it to `<the latest timestamp>` when it is done. The thread/process serving the page checks whether its copy in memory is older than the newest file, if it is and if the newest file is named `pending`,  waits for it to be gone and then loads the data from the the newest file. Otherwise it serves the copy in memory. The `rename` syscall is atomic, so the approach should be free of race conditions. Unlike the approach using lock, it works across processes. It should work fine even for high request rate because most processes/threads will serve the latest data from memory (most of the time).

Reply:
Currently the app is intended to be run as a single node, single process server (the traffic was supposed to be low and no scaling needed).
Do we aim at scaling the app to run as multiple processes?
The requirement was to run the app from within a container, should we prepare the app to be managed by a container orchestration utility like kubernetes with shared volume for data cache?

Reply:
Because of the nature of wsgi servers, I'm not sure if we should allow wsgi server to spawn multiple processes inside a container or run multiple containers, each running single python process...?

Reply:
Customers may want to run multiple instances of the app for high-availability. We can keep individual instances entirely independent (each with it's own cache). The only downside is slightly higher API access rate, but that should be largely mitigated once we make cache updates on-demand. We can declare that wsgi must be configured to fork only 1 process (in once container). We will have to make it clear to customers that if they run multiple instances, they will see increased API access rate.